source/common/upstream/ring_hash_lb.cc

```diff
@@ -37,41 +34,47 @@ void RingHashLoadBalancer::initialize() {
 
 HostConstSharedPtr
 RingHashLoadBalancer::LoadBalancerImpl::chooseHost(LoadBalancerContext* context) {
-  if (global_panic_) {
+  // Make sure we correctly return nullptr for any early chooseHost() calls.
+  if (per_priority_state_.size() == 0) {
+    return nullptr;
+  }
+  // If there is no hash in the context, just choose a random value (this effectively becomes
+  // the random LB but it won't crash if someone configures it this way).
+  // computeHashKey() may be computed on demand, so get it only once.
+  Optional<uint64_t> hash;
+  if (context) {
+    hash = context->computeHashKey();
+  }
+  const uint64_t h = hash.valid() ? hash.value() : random_.random();
+
+  uint32_t priority = LoadBalancerBase::choosePriority(h, per_priority_load_);
+  if (per_priority_state_[priority]->global_panic_) {
     stats_.lb_healthy_panic_.inc();
   }
-  return ring_->chooseHost(context, random_);
+  RingConstSharedPtr ring = per_priority_state_[priority]->current_ring_;
```

Comment:
the compiler will probably optimize this, but I would avoid the local variable so we don't inc/decref the shared pointer.